var/spack/repos/builtin/packages/py-pyopengl/package.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import sys
+
+from spack.package import *
+
+
+class PyPyopengl(PythonPackage):
+    """PyOpenGL is the most common cross platform Python binding to OpenGL and related APIs."""
+
+    homepage = "http://pyopengl.sourceforge.net"
+    url = "https://files.pythonhosted.org/packages/py3/p/pyopengl/PyOpenGL-3.1.6-py3-none-any.whl"
+    list_url = "https://pypi.org/simple/pyopengl/"
+
+    version(
+        "3.1.6",
+        sha256="a7139bc3e15d656feae1f7e3ef68c799941ed43fadc78177a23db7e946c20738",
+        expand=False,
+    )
+
+    variant(
+        "gl",
+        default="glx" if sys.platform.startswith("linux") else "other",
+        values=("glx", "osmesa", "other"),
+        multi=False,
+        description="The OpenGL provider to use",
+    )
+    variant("glu", default=True)
+    variant("glut", default=True)
+
+    conflicts("osmesa", when="gl=glx")
+    conflicts("osmesa", when="gl=other")
+    conflicts("glx", when="gl=osmesa")
+    conflicts("glx", when="gl=other")
+
+    depends_on("python@3.6:", type=("build", "run"))
+    # actually installing PyOpenGL itself just requires python
+    # but tests (and possibly dependent packages) need OpenGL libraries
+    depends_on("gl", type=("build", "run"))
+    depends_on("glx", when="gl=glx", type=("build", "run"))
+    depends_on("osmesa", when="gl=osmesa", type=("build", "run"))
+    depends_on("glu", when="+glu", type=("build", "run"))
+    depends_on("freeglut", when="+glut", type=("build", "run"))
+
+    def setup_build_environment(self, env):
+        # PyOpenGL uses ctypes.cdll (or similar), which searches LD_LIBRARY_PATH
+        lib_dirs = self.spec["gl"].libs.directories
+        if "gl=glx" in self.spec:
+            lib_dirs.extend(self.spec["glx"].libs.directories)
+            env.set("PYOPENGL_PLATFORM", "glx")
+        if "gl=osmesa" in self.spec:
+            lib_dirs.extend(self.spec["osmesa"].libs.directories)
+            env.set("PYOPENGL_PLATFORM", "osmesa")
+        if "+glu" in self.spec:
+            lib_dirs.extend(self.spec["glu"].libs.directories)
+        if "+glut" in self.spec:
+            lib_dirs.append(self.spec["freeglut"].prefix.lib)
```

Comment:
```suggestion
            lib_dirs.append(self.spec["freeglut"].libs.directories[0])
```
If this doesn't find anything, you may need to add a libs property to the package

Reply:
This doesn't work for me. I get the following error even though the libraries are there:
```console
==> Error: NoLibrariesError: Unable to recursively locate freeglut libraries in /home/melven/spack/opt/spack/linux-ubuntu22.04-sandybridge/gcc-11.4.0/freeglut-3.2.2-l4y55y4dgi7jk3puwzfbi3uzyheqpbrj

/home/melven/spack/var/spack/repos/builtin/packages/py-pyopengl/package.py:60, in setup_build_environment:
         57        if "+glu" in self.spec:
         58            lib_dirs.extend(self.spec["glu"].libs.directories)
         59        if "+glut" in self.spec:
  >>     60            lib_dirs.append(self.spec["freeglut"].libs.directories)
         61        libs = ":".join(lib_dirs)
         62        if sys.platform == "darwin":
         63            env.prepend_path("DYLD_FALLBACK_LIBRARY_PATH", libs)


melven@bunti:~/spack$ spack edit py-pyopengl
melven@bunti:~/spack$ ls /home/melven/spack/opt/spack/linux-ubuntu22.04-sandybridge/gcc-11.4.0/freeglut-3.2.2-l4y55y4dgi7jk3puwzfbi3uzyheqpbrj
include  lib
melven@bunti:~/spack$ ls /home/melven/spack/opt/spack/linux-ubuntu22.04-sandybridge/gcc-11.4.0/freeglut-3.2.2-l4y55y4dgi7jk3puwzfbi3uzyheqpbrj/lib/
cmake  libglut.so  libglut.so.3  libglut.so.3.11.1  pkgconfig
```

Reply:
Ok, that's a problem of freeglut which provides a library named libglut but does not set the `libs` property in the package.py.

I added that in  #39300. With this change, also `spec["freeglut"].libs.directories` works correctly.

Reply:
This MR depends on #39300 now!